content/blogposts/2022/ranking-in-a-week.md

```diff
@@ -0,0 +1,120 @@
+
+First, we output `documents.csv`, which correlates a repo and file path pair with a randomly generated identifier (increasing integers for the sake of simplicity in this example). These items correspond to the white vertices in the graph of text documents shown earlier in this post.
+
+```csv
+pathID,repo,path
+1,github.com/sourcegraph/blog-post-example,cmd/app/main.go
+2,github.com/sourcegraph/blog-post-example,cmd/audit/main.go
+3,github.com/sourcegraph/blog-post-example,internal/database.go
+4,github.com/sourcegraph/blog-post-example,internal/experiments.go
+5,github.com/sourcegraph/blog-post-example,internal/handler.go
+6,github.com/sourcegraph/blog-post-example,internal/init.go
+```
+
+Next, we output `references.csv`, which corresponds to the edges _between_ the white vertices. Such an edge is output when there is a direct reference from a variable defined in another file.
+
+```csv
+from,to
+1,3
+1,4
+2,5
+3,4
+4,3
+4,5
+5,3
+5,4
+6,3
+6,5
+```
+
+To support references between repositories, we output `monikers.csv`, which will declare a set of _monikers_ that are either declared in or referenced by a particular path. If one repository declares an `import` moniker, and another repository declares an `export` moniker with the same scheme and identifier, then we create an edge between the documents across repositories.
+
+```csv
+pathID,monikerType,schema,identifier
+1,import,gomod,github.com/sourcegraph/log/Logger
+1,import,gomod,github.com/sourcegraph/log/Logger.Log
+5,import,gomod,github.com/sourcegraph/log/Logger
+5,import,gomod,github.com/sourcegraph/log/Logger.Log
+6,import,gomod,github.com/sourcegraph/log/Logger
+6,import,gomod,github.com/sourcegraph/log/Logger.Log
+```
+
+We would later round out this implemetation by also [pruning irrelevant data](https://github.com/sourcegraph/sourcegraph/pull/43832) from the target GCS bucket as newer indexes for repositories replace older indexes. Since the landing of this export task, it has become the basis of several other solutions currently in-flight (details to come in future blog posts).
+
+On October 27th, after a few successful runs of a distributed Spark job over the graph described by these CSV files and hand-vetting the document scores results to ensure our mental model and reality match, we begin to [consume PageRank outputs into our database](https://github.com/sourcegraph/sourcegraph/pull/43566), for use in the stub service implementation referenced earlier. This "completes the loop" of data from the Sourcegraph instance's point of view: we are continuously refreshing our output graph, and continuously re-consuming PageRanks cores as we re-run the computation.
```

Comment:
Not sure if we want to go into these details, but I think it was interesting that we ended up using bidirectional edges because it produced results that captured more accurate our mental model of what "authorative files" look like. When you have the default direction, PageRank assigns low scores to `main`-like files that depend on a lot of other code, but have no inbound edges.

Reply:
```suggestion
By October 27th, we had a few successful runs of a distributed Spark job over the graph described by these CSV files and hand-vetted the document scores results to ensure our mental model and reality match. In order to build a better intuition of how PageRank will relate to source code documents, we constructed several variations of the graph:

1. Edges are directional, and `a -> b` exists when `a` refers to a symbol defined in `b`
1. Edges are directional, and `b -> a` exists when `a` refers to a symbol defined in `b`
1. Edges are undirected, and both `a -> b` and `b -> a` exist one refers to a symbol defined in the other

Interestingly, we ended up running PageRank over an undirected graph, as directional edges tended to rank auto-generated files with _tons_ of definitions (e.g., a protobuf code gen) very highly, but rank `main`-like files that refers to lots of definitions but defines few themselves very low. This is the opposite of what we want: the glue code with few inbound edges but lots of (unique) outbound edges indicates the collision of two program boundaries. These usages are precisely the ones that should be advertised.

We could then begin to [consume PageRank outputs into our database](https://github.com/sourcegraph/sourcegraph/pull/43566), for use in the stub service implementation referenced earlier. This "completes the loop" of data from the Sourcegraph instance's point of view: we are continuously refreshing our output graph, and continuously re-consuming PageRanks cores as we re-run the computation.
```